scriptworker/task.py

```diff
@@ -490,8 +514,38 @@ def prepare_to_run_task(context, claim_task):
             raise
 
 
-# kill {{{1
-async def kill(pid, sleep_time=1):
+# max_timeout {{{1
+async def max_timeout(context, proc, timeout):
+    """Kill the task if it takes longer than ``timeout`` seconds.
+
+    Args:
+        context (scriptworker.context.Context): the scriptworker context.
+        proc (subprocess.Process): the subprocess proc. This is compared against
+            context.proc to make sure we're killing the right task.
+        timeout (int): the timeout, in seconds, for the task.
+
+    Raises:
+        ScriptWorkerTaskException, if the task is killed.
+
+    Returns:
+        None, if the task finishes before the timeout.
+
+    """
+    await asyncio.sleep(timeout)
+    if proc is not context.proc:
+        # Since this coroutine is run in parallel to the task, it's possible
+        # the running task is no longer the task we were assigned to monitor
+        # for timeout. If `context.proc` doesn't match `proc`, bail.
```

Comment:
Thanks. This is much clearer. Now that we won't hit the next task, is there a reason we need to bail? I can see an argument about wanting to avoid hitting something due to PID reuse. I'd be happy for this to be captured in an issue and postponed.

Reply:
Now that I'm canceling the timeout task at the end of the loop, I think we might be ok not bailing. However,

- yes, we might kill a reused pid,
- we're executing code that shouldn't be needed, slowing things unnecessarily
- the timeout cancel is new and untested, so it's possible it's fallible

It might make sense, though.

Reply:
Oh, and we might kill the existing task if we raise an exception mid-run. That might not happen if we don't retrieve the future's exception, just another piece in the not-sure column.